### PR TITLE
Simplify validation test setup

### PR DESCRIPTION
**Pull request by marchof**
### Current Setup

For each validation test we currently have two files

* `SzenarioXTest.java` and
* `SzenarioXTarget.java`

where the target file uses a special marker syntax on certain lines. These markers are then referred to in the test case and assertions for this lines are executed. For example in the target file we have

    if (f() | f()) { // $line-orFF$

which is later used in the test case:

    assertLine("orFF", ICounter.FULLY_COVERED, 1, 1);

The link between test targets and tests via ids makes test maintainance cumbersome.

### Proposed Solution

The assertions could also directly be included in the target file with a simple statement syntax which is similar to Java. For example:

    if (f() | f()) { // > assertFullyCovered(1, 1);

The expressions refer to methods which are available at the corresponding test case. When the method is called the line number is inserted as the first parameter. While `ValidationTestBase` has a couple of predefined methods test cases can add more assertion methods e.g. for compiler version specific assertions.

A unique marker (e.g. `>` in the example above) is required to distinguish executable assertions from regular comments.

### Work Log

- [x] Hack prototype for new validation interpreter
- [x] Verify feasibility on complex cases like `FinallyTest`
- [x] Convert all tests to new syntax
- [x] Remove $line-tags$ support
- [x] Clean-up `ValidationTestBase` implementation

### CODE REVIEW

**Comment by marchof** [comment]
@Godin Can you please have a look? From my point of view the refactoring is complete. I intentionally left the single commits to make my steps more transparent for the review. Of course we will squash before merge.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+		final List<Object> args = new ArrayList<Object>();
+		expect('(');
+		if (!accept(')')) {
+			args.add(argument());
+			while (!accept(')')) {
+				expect(',');
+				args.add(argument());
+			}
+		}
+		expect(';');
```
Comment (Godin) [inline org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java]: To me seems that removal of `;` is safe. This will make tests cleaner. After replacement on
```
		if (tokenizer.nextToken() != ';') {
			tokenizer.pushBack();
		}
```
all tests except `StatementParserTest.should_fail_when_semicolon_is_missing` pass.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java [hunk 2]
```diff
@@ -12,92 +12,161 @@
...
 public class Source {
 
-	private static final Pattern TAG_PATTERN = Pattern
-			.compile("\\$line-(.*)\\$");
+	private static final String SRC_LOCATION = "src/";
+
+	private static final String SRC_ENCODING = "UTF-8";
+
+	private static final Pattern EXEC_COMMENT_PATTERN = Pattern
+			.compile("//\\s*>(.*)");
```
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java]: I like omitting `>` as it is less verbose - but way harder for the parser: We have `//` e.g. in multi-line comments referencing a URL. Our current naive approach matches single lines only. We cannot detect whether we're within a multi-line comment.

Any ideas?

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java [hunk 3]
```diff
@@ -88,79 +84,85 @@ private void analyze(final ExecutionDataStore store) throws IOException {
...
 	}
 
-	protected final Source getSource() {
-		return source;
+	/*
+	 * Predefined assertion methods:
+	 */
+
+	private void assertCoverage(final Line line, final int insnStatus,
+			final int mb, final int cb) {
```
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java]: I used this in analogy to our XML scheme to have shorter method declarations. If you prefer I can reword it to `missedBranches` and `coveredBranches`. Probably the better choice.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java (hunk 1 above)
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java]: My intention was to use Java syntax. What about multiple invocations in a single comment? Should we require a colon here as a separator?

File: org.jacoco.core.test.validation.java5/src/org/jacoco/core/test/validation/java5/FinallyTest.java [hunk 4]
```diff
@@ -35,180 +40,101 @@
...
-	@Test
-	public void example() {
+	@Before
+	@Override
+	public void setup() throws Exception {
+		super.setup();
+		tags = new HashMap<Integer, String>();
+	}
+
+	public void assertFinallyNormalExecution(final Line line) {
```
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test.validation.java5/src/org/jacoco/core/test/validation/java5/FinallyTest.java]: ✔️ Done.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/JavaStatementExecutor.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ *    
+ *******************************************************************************/
+package org.jacoco.core.test.validation;
+
+import java.lang.reflect.InvocationTargetException;
+
+/**
+ * Executes statements against a given Java object instance.
+ */
+class JavaStatementExecutor implements StatementParser.IStatementVisitor {
```
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/JavaStatementExecutor.java]: ✔️ ✔️ Both done.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/JavaStatementExecutor.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		System.arraycopy(b, 0, result, a.length, b.length);
+		return result;
+	}
+
+	private static Class<?>[] getTypes(Object[] instances) {
+		final Class<?>[] classes = new Class[instances.length];
+		for (int i = 0; i < instances.length; i++) {
+			Class<? extends Object> c = instances[i].getClass();
+			if (c == Integer.class) {
+				// We always assume primitive int parameters:
```
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/JavaStatementExecutor.java]: ✔️ Done.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java [hunk 7]
```diff
@@ -88,79 +84,85 @@ private void analyze(final ExecutionDataStore store) throws IOException {
...
-		throw new AssertionError();
 	}
 
-	protected final Source getSource() {
-		return source;
+	/*
+	 * Predefined assertion methods:
+	 */
+
+	private void assertCoverage(final Line line, final int insnStatus,
```
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java]: Would you prefer to have them static?

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java]: @marchof of course you right that in general case real lexer is required, and I'm ashamed that missed/forgot this fact. And in general case lexer of source language - multiline string literals in Kotlin/Groovy, etc.

However in our case seems that the only problem is URLs. What do you think about regex with negative lookbehind to exclude them - `(?<!https?:)//(.*)` ?

My quick attempt

```diff
--- org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java
+++ org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java
@@ -45,7 +45,7 @@
 	private static final String SRC_ENCODING = "UTF-8";
 
 	private static final Pattern EXEC_COMMENT_PATTERN = Pattern
-			.compile("//\\s*>(.*)");
+			.compile("(?<!https?:)//(.*)");
 
 	/**
 	 * Represents a single line in a source file.
--- org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java
+++ org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java
@@ -73,6 +73,7 @@
 		tokenizer = new StreamTokenizer(new StringReader(source));
 		tokenizer.resetSyntax();
 		tokenizer.whitespaceChars(' ', ' ');
+		tokenizer.whitespaceChars('>', '>');
 		tokenizer.whitespaceChars('\t', '\t');
 		tokenizer.wordChars('a', 'z');
 		tokenizer.wordChars('A', 'Z');
```

and few conversions of comments from `//` into `/*`, shows that this is sufficient.

And so far I can imagine only one highly unlikely corner-case when this can miss executable comment

```
	enum E {
		http
	}

	static void example(E e) {
		switch (e) {
		case http:// executable comment
		}
	}
```

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java (hunk 1 above)
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java]: Colon (`:`) sounds/looks even more weird than current presence of semicolon.

Semicolon IMO really redundant in case of single invocation. And why not omit separator in case of multiple invocations? Still will be Java-like syntax

```diff
--- org.jacoco.core.test.validation.java5/src/org/jacoco/core/test/validation/java5/targets/FinallyTarget.java
+++ org.jacoco.core.test.validation.java5/src/org/jacoco/core/test/validation/java5/targets/FinallyTarget.java
@@ -75,11 +75,11 @@
 	private static void catchNotExecuted() {
 		try {
 			nop("try");
-		} catch (Exception e) { // > tag("catchNotExecuted");
-			nop("catch"); // > assertNotCovered();
-		} finally { // > assertEmpty();
-			nop("finally"); // >assertFullyCovered();tag("catchNotExecuted.1");
-		} // > assertEmpty(); tag("catchNotExecuted.2");
+		} catch (Exception e) { // tag("catchNotExecuted")
+			nop("catch"); // assertNotCovered()
+		} finally { // assertEmpty()
+			nop("finally"); // assertFullyCovered() tag("catchNotExecuted.1")
+		} // assertEmpty() tag("catchNotExecuted.2")
 	}
 
 	private static void emptyCatch() {
```

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java]: For the potential external contributors, I think `missedBranches` and `coveredBranches` are better even if XML uses `mb` and `cb`.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java]: don't know - this was just observation

[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java]: ✔️ No action required.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java]: ✔️ Done.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java (hunk 1 above)
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java]: @Godin Sorry, I wanted to say *semicolon*.

Here is another catch: For long lines we must avoid whitespaces, otherwise the auto-formatter wraps the commend into multiple lines. In this case the comment would read:

    nop("finally"); // assertFullyCovered()tag("catchNotExecuted.1")

If this is ok for you I would remove semicolon from the syntax at all.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java]: Ok, I'll give it a try.

[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java]: ✔️ Done. Actually identified 2 instances, where I forgot `>`. So less errorprone now.

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java [hunk 8]
```diff
@@ -12,92 +12,161 @@
...
+		}
+
+		public ILine getCoverage() {
+			return coverage;
+		}
+
+		/**
+		 * @return exectable comment if present or <code>null</code>
+		 */
+		public String getExecutableComment() {
```
[the remark this answers is not in the document]
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java]: As it is now a generic `getComment()` I left it here. But I moved the comment about executable assertions to `ValidationTestBase` in 016acb2. Ok?

[the remark this answers is not in the document]
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java]: 👍 ✔️

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java]: which means that I also overlooked them 😉 Now I see - one is `assertFinally` in `Exceptions` and still wondering where is another one? 😆 So definitely less error prone 👍

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java (hunk 1 above)
Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java]: ✔️ Semicolons are gone.

Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java]: BTW, I found two instances where I already forgot to include `>` 😉

**Comment by marchof** [comment]
@Godin All done for now. Do you want me to push a squashed commit or is Github's magic button o.k. for you?

File: org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java (hunk 1 above)
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java]: I would not say that semicolon of a much help for formatting, so 👍 about removal.

About forgotten `>` see https://github.com/jacoco/jacoco/pull/718#discussion_r209002552 - I still don't know how you found second one 😆

**Comment by Godin** [comment]
@marchof I have an additional idea: construction of stack trace allows to have easy navigation in IDEs - straight to the line of comment. Here is PoC:

```diff
--- org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java
+++ org.jacoco.core.test/src/org/jacoco/core/test/validation/ValidationTestBase.java
@@ -124,6 +124,18 @@
 		String msg = String.format("Instructions in %s: %s", line,
 				line.getText());
 		final int actualStatus = coverage.getInstructionCounter().getStatus();
+
+		if (insnStatus != actualStatus) {
+			AssertionError e = new AssertionError(msg);
+			String declaringClass = source.getCoverage().getPackageName().replace('/', '.')
+					+ "." + source.getCoverage().getName().replace(".java", "");
+			String fileName = source.getCoverage().getName();
+			e.setStackTrace(
+					new StackTraceElement[] { new StackTraceElement(declaringClass,
+							"foo", fileName, line.getNr()) });
+			throw e;
+		}
+
 		assertEquals(msg, STATUS_NAME[insnStatus], STATUS_NAME[actualStatus]);
 
 		msg = String.format("Branches in %s: %s", line, line.getText());
```

not too crazy? WDYT?

**Comment by marchof** [comment]
👍 

AFAIK at least for Eclipse adding `(Target.java:123)` to the message would be enough. No need to tweak the stack trace.

**Comment by marchof** [comment]
@Godin Ah, now I see: It does not show up in the Console, only in JUnit view. Here a stack trace is required.

**Comment by Godin** [comment]
@marchof fine for me, because this is enough for IntelliJ, which shows message in console and so I'm one-click away from location. I was more thinking about you - is it fine that in case of Eclipse you need additional click on "Show Stack Trace in Console View" instead of one in "Failure Trace"?

eclipse
